packages/dds/tree/src/feature-libraries/modular-schema/discrepancies.ts

```diff
@@ -0,0 +1,339 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	type FieldKey,
+	type FieldKindIdentifier,
+	LeafNodeStoredSchema,
+	MapNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	type TreeFieldStoredSchema,
+	type TreeNodeSchemaIdentifier,
+	type TreeStoredSchema,
+	type TreeTypeSet,
+	type ValueSchema,
+} from "../../core/index.js";
+
+interface AllowedTypeIncompatibility {
+	identifier: string | undefined; // undefined indicates root field schema
+	mismatch: "allowedTypes";
+	/**
+	 * List of allowed type identifiers in viewed schema
+	 */
+	view: string[];
+	/**
+	 * List of allowed type identifiers in stored schema
+	 */
+	stored: string[];
+}
+
+interface FieldKindIncompatibility {
+	identifier: string | undefined; // undefined indicates root field schema
+	mismatch: "fieldKind";
+	view: FieldKindIdentifier | undefined;
+	stored: FieldKindIdentifier | undefined;
+}
+
+interface ValueSchemaIncompatibility {
+	identifier: string;
+	mismatch: "valueSchema";
+	view: ValueSchema | undefined;
+	stored: ValueSchema | undefined;
+}
+
+type FieldIncompatibility =
+	| AllowedTypeIncompatibility
+	| FieldKindIncompatibility
+	| ValueSchemaIncompatibility;
+
+type SchemaFactoryNodeKind = "object" | "leaf" | "map";
+
+interface NodeKindIncompatibility {
+	identifier: string;
+	mismatch: "nodeKind";
+	view: SchemaFactoryNodeKind | undefined;
+	stored: SchemaFactoryNodeKind | undefined;
+}
+
+interface NodeFieldsIncompatibility {
+	identifier: string;
+	mismatch: "fields";
+	differences: FieldIncompatibility[];
+}
+
+type NodeIncompatibility = NodeKindIncompatibility | NodeFieldsIncompatibility;
+
+type Incompatibility = FieldIncompatibility | NodeIncompatibility;
+
+/**
+ * The workflow for finding schema incompatibilities:
+ * 1. Identify discrepancies in the root schema.
+ * 2. For each node schema in the `view`:
+ * - Check if the node schema exists in the `stored`.
+ * - If it exists in the `stored`, perform exhaustive validation to identify all discrepancies.
+ * 3. For each node schema in the `stored`, check if it exists in the `view`. The overlapping parts
+ * were already tracked in the previous step.
+ *
+ * @returns the discrepancies between two TreeStoredSchema objects
+ */
+export function getAllowedContentIncompatibilities(
+	view: TreeStoredSchema,
+	stored: TreeStoredSchema,
+): Incompatibility[] {
```

Comment:
Could you file some work to convert this to a generator? It looks like it should be pretty straightforward to do, and will allow us to be more efficient in detecting incompatible schema (it will enable shortcutting / early exit like we do today). It should be a mechanical change though so fine to do in a follow-up.